### DISCUSSION TITLE
How to return numpy ndarray objects from Rust to Python?

### DISCUSSION THREAD

**Discussion by ifsheldon** [PyO3/pyo3#2042] [Questions] [answered]
Hi everyone! I'm new to pyo3 and I'd like to ask how I can return numpy ndarrays from Rust to Python. 

Simply put, I want to write a function whose input and output are numpy ndarrays, write it in Rust with Rust-native data types and glue it to Python. I've been reading the documentations of pyo3 and rust-numpy, but I didn't find out related contents. 

I know `Vec<T>` in Rust can be converted to `list[T]` in Python, but I don't know how much overhead it has to convert `Vec<T>` to `list[T]` and then to `np.ndarray`. If there's no direct mapping from `np.ndarray` to Rust's data representation, I think we will have to do something like `np.ndarray` -> `list[T]` -> `Vec<T>` -> Process in Rust -> `Vec<T>` -> `list<T>` -> `np.ndarray`.

If you have done that before or you know where the related documentation is, please give me a pointer. Thanks!

**Comment by davidhewitt** [comment, marked as answer]
You're probably looking for [rust-numpy](https://github.com/PyO3/rust-numpy). See also #2029

**Reply by ifsheldon** [reply]
OK, I found the example in rust-numpy, thank you!
